dlib/test/dnn.cpp

```diff
@@ -2495,6 +2495,113 @@ namespace
         DLIB_TEST_MSG(error_after < 1e-6, "Autoencoder error after training = " << error_after);
     }
 
+// ----------------------------------------------------------------------------------------
+
+    void test_loss_mean_squared_per_channel_and_pixel()
+    {
+        print_spinner();
+
+        dlib::rand rnd;
+        const long num_channels = 9;
+        const int num_rows = 8;
+        const int num_cols = 8;
+        const int num_samples = 1000;
+
+        // In this test, we are going to generate an image black image with 9
+        // white pixels and set up a network to output each pixel in a separate
+        // channel. The pixel numbering goes as follows (we assume padding at
+        // the borders and between pixels:
+        //
+        //  -------      -------
+        // | * * * |    | 0 1 2 |
+        // | * * * |    | 3 4 5 |
+        // | * * * |    | 6 7 8 |
+        //  -------      -------
+        //
+        // the main thing to notice when using this layer is that the output of
+        // the subnetwork needs to have the same number of channels as the
+        // channels we specify
+        using net_type = loss_mean_squared_per_channel_and_pixel<num_channels,
+                            htan<con<num_channels, 1, 1, 1, 1,
+                            relu<cont<32, 2, 2, 2, 2,
+                            relu<con<32, 2, 2, 2, 2,
+                            input<matrix<float>>>>>>>>>;
+        net_type net;
+        matrix<float> background = zeros_matrix<float>(num_rows, num_cols);
```

Comment:
In order to make sure the loss decreases, I've made this small change:
``` cpp
matrix<float> background = -ones_matrix<float>(num_rows, num_cols);
```
Since the network output is a `htan` the range of its values are `[-1, 1]`. That also has the benefit of making the network have a higher loss at the beginning:
- `error_before = 1.0636`
- `error_after = 0.0370037`

If the problem was the network not reducing the loss enough, would that solve it?
Or do you still prefer your example with random linear functions?
The reason I kind of like this example, is because it shows how it can be used to learn a network that estimates landmark heatmaps.